### DISCUSSION TITLE
Global vs Local events / transitions

### DISCUSSION THREAD

**Discussion by adammartiska** [statelyai/xstate#2700] [General] [answered]
Hello,
I am solving issue with redirecting user to another page using browserHistory functions. If there are any unsaved data, I want to ask user at first if he really wants to leave the page with dialog that is invoked by a service from xstate. I am using effect when user wants to navigate away: 
```
  React.useEffect(() => {
    return history.block((location) => {
      if (areTestPointParametersDirty) {
        service.send({ type: 'ROUTE_CHANGE_INITIATED', url: location.pathname })
        return false
      }
      return true
    })
  }, [service, areTestPointParametersDirty])
```

Upon this event arrives into our main machine, I want to start some services and calculation ( for example save the data on the server ) and then inform my child machine, that some of its data has been updated. However when I registered global event listener ROUTE_CHANGE_INITIATED , if I want to send some data to child machine I am getting error: 
**Unable to send event to child 'MCTMachine' from service 'root'.** What I have noticed that when I register this event in some of the states that I am currently in - it is working as it should - communication between parent and child machine is working fine. I don't know why this happens if there is any difference how the machine behaves on GLOBAL / LOCAL events. In my research I have found out that **history.block()** is unmounting component - that's why IMO I can not communicate with child machine ( but what my head can not understand is why I can communicate from for example **Idle** state ).

**Comment by davidkpiano** [comment]
If the component is being unmounted, the local machine that is interpreted in that component is _stopped_. 

Can you share a full reproducible code example?

**Reply by adammartiska** [reply]
We are using machine as hooks this is piece of code of our main and child machine that is IMO important.
**MAIN MACHINE :** 
```
const createMainMachine = ({ parametersMachine }) =>
  createMachine(
    {
      id: 'root',
      type: 'parallel',    
      preserveActionOrder: true,
      context: { ... },
      states: { 
            PageState: { Idle, ReadyForMeasurement, ...},
            TestStrategy: { QuickMeasurement: {}, MCT (THIS IS OUR CHILD MACHINE THAT I WANT TO SEND EVENTS TO): {}       
                  },
       ),

export function useMainMachine({ validateForm } = {}) {
  const parametersMachine = useParametersMachine()
  const [state, sendImperative, service] = useMachine(
    () => createMainMachine({ parametersMachine }),
    {
      actions: useMainMachineActions(),
      services: useMainMachineServices({ validateForm }),
      guards: useMainMachineGuards(),
      devTools: getMode() !== 'production',
    },

  )
  return [state, sendImperative, service]
}

```

This is our **CHILD MCT Machine**:

```
export const MCTMachine = createMachine({
  id: 'mct',
  context: { ... }
    refs: {
      parametersMachine: null,
    },
  },
  type: 'parallel',
  states: {
    Operation: {
      initial: 'Initializing',
      states: {  ... }
       on: {
                **TEST_POINT_UPDATED** => this is event, which I want to listen on in this machine
}
}
)
```

And we have measurement page which is calling main machine as : 
`  const [state, send, service] = useMainMachine({ validateForm })`
And component on measurement page, which is calling MCT child machine : 
```
function MCTTab({ handleChangedTestStrategy }) {
  const _ = useTranslate(messages)

  const mainService = React.useContext(MachineContext)
  const MCTService = useSelector(mainService, selectMCTMachine)
  const [, send] = useService(MCTService)
}
```

**Comment by adammartiska** [comment, marked as answer]
Just found out that all I needed to do registrer global event as internal, my bad. Thank you for help :)
